File: scala/79.scala
object Solution {
    def exist(board: Array[Array[Char]], word: String): Boolean = {
        def search(i: Int, j: Int, k: Int): Boolean = {
            if (k >= word.length()) return true
            (i, j) match {
                case (i,j) if i==board.length || j == board.head.length || i == -1 || j == -1 => false   
                case _ => {
                    if (board(i)(j) != word(k)) false
                    else {
                        val tmp = board(i)(j)
                        board(i)(j) = '0'
                        val res = search(i-1, j, k+1) || search(i+1, j, k+1) || search(i, j-1, k+1) || search(i, j+1, k+1)
                        board(i)(j) = tmp
                        return res
                    }
                }
            }
        }       
        for (i <- 0 until board.length; j <- 0 until board.head.length)
            if (search(i, j, 0)) return true
        false
    }
}

File: leetcodeJava/79.java
package leetcodeJava;

class Solution {
    public boolean exist(char[][] board, String word) {
        for (int i = 0; i < board.length; i++)
            for (int j = 0; j < board[0].length; j++) 
                if (dfs(board, word, i, j)) return true;
        return false;
    }
    
    private boolean dfs(char[][] board, String word, int i, int j) {
        if (word.length() == 0) return true;
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length) return false;
        if (board[i][j] != word.charAt(0)) {
            return false;
        }
        var tmp = board[i][j];
        board[i][j] = ' ';
        var subWord = word.substring(1, word.length());
        if (dfs(board, subWord, i+1, j)) return true;
        if (dfs(board, subWord, i, j+1)) return true;
        if (dfs(board, subWord, i-1, j)) return true;
        if (dfs(board, subWord, i, j-1)) return true;
        board[i][j] = tmp;
        return false;
    }
}
